kafkatail/kafkatail.go

```diff
@@ -2,63 +2,90 @@ package kafkatail
 
 import (
 	"context"
+	"github.com/Shopify/sarama"
+	"github.com/cenk/backoff"
+	"github.com/rubyist/circuitbreaker"
 	"log"
 	"time"
-
-	"github.com/Shopify/sarama"
 )
 
 type Options struct {
 	Server    string `long:"server" description:"kafka server" default:"localhost"`
 	Port      string `long:"port" description:"kafka port" default:"9092"`
 	Topic     string `long:"topic" description:"kafka topic" default:"my_topic"`
 	Partition int32  `long:"partition" description:"partition to read from"`
+	Offset    int64  `long:"offset" description:"offset to start from" default:"sarama.OffsetNewest"`
 }
 
 // GetChans returns a list of channels but it only ever has one entry - the
 // partition on which we're listening.
 // TODO listen on multiple channels to multiple partitions
-func GetChans(ctx context.Context, options Options) ([]chan string, error) {
-	linesChans := make([]chan string, 1, 1)
+func GetChans(ctx context.Context, options Options) ([]chan string, chan error) {
+	linesCh := make([]chan string, 1, 1)
 	lines := make(chan string, 1)
-	linesChans[0] = lines
+	linesCh[0] = lines
+
+	config := sarama.NewConfig()
+	config.Consumer.Return.Errors = true
+
+	brokers := []string{options.Server + ":" + options.Port}
 
-	serverString := options.Server + ":" + options.Port
-	// TODO use a reasonable kafka *Config instead of nil for the new consumer
-	consumer, err := sarama.NewConsumer([]string{serverString}, nil)
+	consumer, err := sarama.NewConsumer(brokers, config)
 	if err != nil {
 		panic(err)
 	}
 
-	partitionConsumer, err := consumer.ConsumePartition(
-		options.Topic, options.Partition, sarama.OffsetNewest)
+	partitionConsumer, err := consumer.ConsumePartition(options.Topic, options.Partition, options.Offset)
 	if err != nil {
 		panic(err)
 	}
 
+	errCh := make(chan error, 1)
+
 	go func() {
-		log.Printf("consumer started")
+		log.Printf("Consumer started for topic %s\n", options.Topic)
+
+		defer func() {
+			log.Printf("Terminating consumer for topic %s\n", options.Topic)
+			close(lines)
+
+			if err := partitionConsumer.Close(); err != nil {
+				log.Fatalln(err)
+			}
+
+			if err := consumer.Close(); err != nil {
+				log.Fatalln(err)
+			}
+		}()
+
+		// TODO: Make these configurable
+		expBackoff := backoff.NewExponentialBackOff()
+		expBackoff.MaxInterval = 30 * time.Second
+		expBackoff.MaxElapsedTime = 5 * time.Minute
+
+		cb := circuit.NewConsecutiveBreaker(10)
+		cb.BackOff = expBackoff
+
 		for {
-			select {
-			case msg := <-partitionConsumer.Messages():
-				if msg != nil {
+			if cb.Ready() {
+				select {
+				case err := <-partitionConsumer.Errors():
+					log.Printf("Error at offset %d in topic %s: %v\n",
+						partitionConsumer.HighWaterMarkOffset()-1, options.Topic, err)
+					cb.Fail()
+
+					if cb.Tripped() && expBackoff.NextBackOff() == backoff.Stop {
```

Comment:
How does this interact with the Sarama's consumer retry? https://godoc.org/github.com/Shopify/sarama#Config

Reply:
Probably not well. It looks like Sarama retries every two seconds by default regardless of what I'm doing above it, so this will likely cause problems. It's a good catch - let me see what I can do about it.